### PR TITLE
Implement specialized filter kernel for `FixedSizeByteArray`

### LINKED ISSUE #6153: Implement `filter` kernel specially for `FixedSizeByteArray`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
We are trying to store UUIDs in arrow Arrays and one obvious thing to do is use `FixedSizeByteArray`

However @samuelcolvin  did some experiments (see https://github.com/apache/datafusion/issues/11170) and found that non obviously using `Decimal128` was actually faster than `FixedSizeByteArray`

One reason for this may be that `FixedSizeByteArray` does not have special case handling in the `filter` kernel 

https://github.com/apache/arrow-rs/blob/f42d2420525a05a9b55461d83b359779ca5cc2a3/arrow-select/src/filter.rs#L320-L383

**Describe the solution you'd like**

1. Add special case code for `FixedSizeByteArray`
2. Add benchmark showing it is faster
3. Add unit tests for functional test coverage

**Describe alternatives you've considered**

**Additional context**
This was pointed out by @samuelcolvin  on  https://github.com/apache/datafusion/issues/11170#issuecomment-2253527396

### PR DESCRIPTION
**Pull request by chloro-pn**
# Which issue does this PR close?

Closes  #6153

### CODE REVIEW

**Comment by alamb** [comment]
> The current issue is that `criterion` [cannot test private members in the module](https://bheisler.github.io/criterion.rs/book/user_guide/known_limitations.html), so I have to set some members in `filter.rs` to public or provide access interfaces, which may not be the best practice.


Perhaps you could add some cases to https://github.com/apache/arrow-rs/blob/master/arrow/benches/filter_kernels.rs rather than a new benchmark? I think that has a pretty good setup for testing

**Comment by alamb** [comment]
(thank you very much, BTW)

**Comment by chloro-pn** [comment]
Okay, get it. I will add the benchmark in the first PR and the refactoring part in the second PR, instead of using `Comparing Functions`

File: arrow-select/src/filter.rs [hunk 1]
```diff
@@ -707,6 +710,62 @@ fn filter_byte_view<T: ByteViewType>(
     GenericByteViewArray::from(unsafe { builder.build_unchecked() })
 }
 
+fn filter_fixed_size_binary(
+    array: &FixedSizeBinaryArray,
+    predicate: &FilterPredicate,
+) -> FixedSizeBinaryArray {
+    let values: &[u8] = array.values();
+    let value_length = array.value_length() as usize;
+    let calcualte_offset_from_index = |index: usize| index * value_length;
```
Comment (alamb) [inline arrow-select/src/filter.rs:719]: There is a typo here:

```suggestion
    let calculate_offset_from_index = |index: usize| index * value_length;
```

File: arrow-buffer/src/buffer/mutable.rs [hunk 2]
```diff
@@ -896,11 +946,19 @@ mod tests {
     #[test]
     fn test_from_trusted_len_iter() {
         let iter = vec![1u32, 2].into_iter();
-        let buf = unsafe { Buffer::from_trusted_len_iter(iter) };
+        let buf = unsafe { MutableBuffer::from_trusted_len_iter(iter) };
```
Comment (alamb) [inline arrow-buffer/src/buffer/mutable.rs:949]: why this change?

File: arrow-buffer/src/buffer/mutable.rs [hunk 3]
```diff
@@ -668,6 +668,56 @@ impl MutableBuffer {
...
+    /// # use arrow_buffer::MutableBuffer;
+    /// # use arrow_buffer::ToByteSlice;
+    /// let iter = vec![[1_u8, 2].to_byte_slice(), [3_u8, 4].to_byte_slice()].into_iter();
+    /// let buf = unsafe {
+    ///     MutableBuffer::from_trusted_len_iter_slice_u8(iter, 2)
+    /// };
+    /// assert_eq!(4, buf.len());
+    /// ```
+    /// # Safety
+    /// This method assumes that the iterator's size is correct and is undefined behavior
```
Comment (alamb) [inline arrow-buffer/src/buffer/mutable.rs:687]: I think it also assumes that each slice is `item_size` bytes

File: arrow-select/src/filter.rs [hunk 4]
```diff
@@ -707,6 +710,62 @@ fn filter_byte_view<T: ByteViewType>(
...
+            }
+            buffer
+        }
+        IterationStrategy::IndexIterator => {
+            let iter = IndexIterator::new(&predicate.filter, predicate.count).map(|x| {
+                &values[calcualte_offset_from_index(x)..calcualte_offset_from_index(x + 1)]
+            });
+
+            // SAFETY: IndexIterator is trusted length
+            unsafe { MutableBuffer::from_trusted_len_iter_slice_u8(iter, value_length) }
```
Comment (alamb) [inline arrow-select/src/filter.rs:745]: I wonder how much better performance does this get than calculating `MutableBuffer::extend_from_slice`? I looked at the implementation of extend_from_slice and it does very much the same thing

File: arrow-select/src/filter.rs [hunk 5]
```diff
@@ -985,6 +1044,25 @@ mod tests {
...
     }
 
+    #[test]
+    fn test_filter_fixed_binary() {
+        let v1 = [1_u8, 2];
+        let v2 = [3_u8, 4];
+        let v3 = [5_u8, 6];
+        let v = vec![&v1, &v2, &v3];
+        let a = FixedSizeBinaryArray::from(v);
+        let b = BooleanArray::from(vec![true, false, true]);
```
Comment (alamb) [inline arrow-select/src/filter.rs:1054]: Given there are different code paths for the different types of filters (aka different patterns in `BooleanArray`) Perhaps we can add some coverage to ensure the other filter patterns are hit.

**Review by alamb — COMMENTED** [review]
Thanks @chloro-pn -- I think this PR is getting quite close

I am a bit worried about the complexity and potential danger of adding `MutableBuffer::from_trusted_len_iter_slice_u8` unless there is some compelling justification for it (aka performance)

I am also going to run some benchmarks on this PR on my own test rig

File: arrow-buffer/src/buffer/mutable.rs (hunk 2 above)
Comment (chloro-pn) [reply arrow-buffer/src/buffer/mutable.rs:949]: This modification is not directly related to this PR.
Because this is the file that implements `MutableBuffer`, I thought it was a typographical error. The tests in this file should mainly focus on `MutableBuffer`.

File: arrow-select/src/filter.rs (hunk 4 above)
Comment (chloro-pn) [reply arrow-select/src/filter.rs:745]: The reason for implementing the `from_trusted_len_iter_slice_u8` method is that it involves performing memory copy operations based on `value_length`:
```
        for item in iterator {
            // note how there is no reserve here (compared with `extend_from_iter`)
            let src = item.to_byte_slice().as_ptr();
            std::ptr::copy_nonoverlapping(src, dst, item_size);
            dst = dst.add(item_size);
        }
```
If we continue to use the `extend_from_slice`, we will have to perform memory copy operations according to the size of `u8`. As far as I know, in many platforms, memory copy operations are optimized through SIMD instructions, so copying multiple bytes at once obviously has better performance.

**Comment by chloro-pn** [comment]
Except for the part I replied that needs further discussion, all other code reviews will be fixed

File: arrow-select/src/filter.rs (hunk 4 above)
Comment (chloro-pn) [reply arrow-select/src/filter.rs:745]: I have not measured the performance improvement brought by using `from_trusted_len_iter_slice_u8`. 
If necessary, I can use two methods to implement the logic here separately and compare the performance differences in the benchmark.

**Comment by alamb** [comment]
I tried to run my benchmark script (https://github.com/alamb/datafusion-benchmarking/blob/main/compare_arrow_branch.sh in case you care) and it didn't report on the new benchmarks you added. I think it is because this branch doesn't have the latest master changes too

SO I merged from main and restarted the tests

**Comment by alamb** [comment]
The benchmark results look good to me:

```rust
filter fsb with value length 20 (kept 1/2)                                    1.00    169.3±0.40µs        ? ?/sec    1.68    284.8±0.82µs        ? ?/sec
filter fsb with value length 20 high selectivity (kept 1023/1024)             1.00     69.5±0.32µs        ? ?/sec    1.06     73.7±2.70µs        ? ?/sec
filter fsb with value length 20 low selectivity (kept 1/1024)                 1.00      3.2±0.01µs        ? ?/sec    1.14      3.6±0.01µs        ? ?/sec
filter fsb with value length 5 (kept 1/2)                                     1.00    169.0±0.30µs        ? ?/sec    1.61    272.6±0.67µs        ? ?/sec
filter fsb with value length 5 high selectivity (kept 1023/1024)              1.00     10.3±0.33µs        ? ?/sec    1.10     11.4±0.63µs        ? ?/sec
filter fsb with value length 5 low selectivity (kept 1/1024)                  1.00      3.2±0.00µs        ? ?/sec    1.15      3.6±0.01µs        ? ?/sec
filter fsb with value length 50 (kept 1/2)                                    1.00    188.9±2.39µs        ? ?/sec    1.73    326.8±2.32µs        ? ?/sec
filter fsb with value length 50 high selectivity (kept 1023/1024)             1.05   242.5±24.39µs        ? ?/sec    1.00   230.0±13.04µs        ? ?/sec
filter fsb with value length 50 low selectivity (kept 1/1024)                 1.00      3.1±0.00µs        ? ?/sec    1.13      3.5±0.01µs        ? ?/sec
```

**Review by alamb — APPROVED** [review]
Thanks again @chloro-pn 

I think this PR could be merged as it follows existing patterns and shows a significant performance improvement. 

I think it would be better if you could:
1. Review the tests and add some additional coverage
2. See if we can avoid introducing `MutableBuffer::from_trusted_len_iter_slice_u8` as discussed in https://github.com/apache/arrow-rs/pull/6178/files#r1705589325 to keep the API surface down.

I think we could also merge this PR and then try removing `from_trusted_len_iter_slice_u8` in a follow on PR and measure the performance difference, if any, as we now have a nice benchmark thanks to you

File: arrow-buffer/src/buffer/mutable.rs [hunk 6]
```diff
@@ -668,6 +668,56 @@ impl MutableBuffer {
...
+    /// to use it on an iterator that reports an incorrect length.
+    // This implementation is required for two reasons:
+    // 1. there is no trait `TrustedLen` in stable rust and therefore
+    //    we can't specialize `extend` for `TrustedLen` like `Vec` does.
+    // 2. `from_trusted_len_iter_slice_u8` is faster.
+    #[inline]
+    pub unsafe fn from_trusted_len_iter_slice_u8<I: Iterator<Item = &'a [u8]>>(
```
Comment (tustvold) [inline arrow-buffer/src/buffer/mutable.rs:694]: It is hard for me to see why this would be any faster than

```
let mut buffer = MutableBuffer::new(len.checked_mul(item_size).unwrap());
iterator.for_each(|item| buffer.push_unchecked(item));
```

File: arrow-select/src/filter.rs (hunk 4 above)
Comment (tustvold) [reply arrow-select/src/filter.rs:745]: As mentioned above, if the performance of `extend_from_slice` is insufficient, which I think would need to be measured as memcpy is not typically inlined and may dominate, you could use `push_unchecked`.

**Review by tustvold — CHANGES_REQUESTED** [review]
I think we either need to justify adding `from_trusted_len_iter_slice_u8` with concrete data, or remove it. Adding subtle unsafe methods is something that should only be done sparingly

File: arrow-buffer/src/buffer/mutable.rs (hunk 6 above)
Comment (chloro-pn) [reply arrow-buffer/src/buffer/mutable.rs:694]: the "faster" here means `"/// Prefer this to collect whenever possible, as it is faster."`.
which appears in the first half of this function's comment.
* from_trusted_len_iter
* from_trusted_len_iter_bool
* from_trusted_len_iter_slice_u8

The logic implemented by these functions is similar, and the comment are also similar.

File: arrow-select/src/filter.rs (hunk 4 above)
Comment (chloro-pn) [reply arrow-select/src/filter.rs:745]:
```
    /// Extends the buffer with a new item, without checking for sufficient capacity
    /// # Safety
    /// Caller must ensure that the capacity()-len()>=`size_of<T>`()
    #[inline]
    pub unsafe fn push_unchecked<T: ToByteSlice>(&mut self, item: T) {
        let additional = std::mem::size_of::<T>();
        let src = item.to_byte_slice().as_ptr();
        let dst = self.data.as_ptr().add(self.len);
        std::ptr::copy_nonoverlapping(src, dst, additional);
        self.len += additional;
    }
```

The implementation of `push_unchecked` also calls `std::ptr::copy_nonoverlapping`, so I don't think this approach can bring better performance.

File: arrow-buffer/src/buffer/mutable.rs (hunk 6 above)
Comment (tustvold) [reply arrow-buffer/src/buffer/mutable.rs:694]: Right from_trusted_len_iter and from_trusted_len_iter_bool make sense because they're using primitives with a fixed width known at compile time, this allows for effective vectorisation. With variable width types, as in from_trusted_len_iter_slice_u8, the compiler will just insert a memcpy for each element, and there will be no performance benefit

File: arrow-buffer/src/buffer/mutable.rs (hunk 3 above)
Comment (chloro-pn) [reply arrow-buffer/src/buffer/mutable.rs:687]: add
`   /// # Panic`
`    /// This method will panic if &[u8].len() != item_size.`

File: arrow-select/src/filter.rs (hunk 5 above)
Comment (chloro-pn) [reply arrow-select/src/filter.rs:1054]: fixed

File: arrow-select/src/filter.rs (hunk 4 above)
Comment (tustvold) [reply arrow-select/src/filter.rs:745]: Right so you agree we don't need from_trusted_len_iter_slice_u8 and can remove it?

**Comment by chloro-pn** [comment]
by the way :
```
    /// Extends the buffer with a new item, without checking for sufficient capacity
    /// # Safety
    /// Caller must ensure that the capacity()-len()>=`size_of<T>`()
    #[inline]
    pub unsafe fn push_unchecked<T: ToByteSlice>(&mut self, item: T) {
        let additional = std::mem::size_of::<T>();
        let src = item.to_byte_slice().as_ptr();
        let dst = self.data.as_ptr().add(self.len);
        std::ptr::copy_nonoverlapping(src, dst, additional);
        self.len += additional;
    }
```
The type T of `item` here cannot be a slice type. If we want to increase the ability of this function to handle slice types, it may involve incompatible modifications.

File: arrow-select/src/filter.rs (hunk 4 above)
Comment (chloro-pn) [reply arrow-select/src/filter.rs:745]: If it can be implemented through `push_unchecked`, then `from_trusted_len_iter_slice_u8` can be removed

Comment (tustvold) [reply arrow-select/src/filter.rs:745]: Aah that's annoying, in which case the next step is to get some numbers comparing against extend_from_slice.

I guess you could use `set_len` and do the copy_non_overlapping manually, but I'd really like to see numbers comparing against extend_from_slice, this smells very strongly of premature optimisation

Comment (chloro-pn) [reply arrow-select/src/filter.rs:745]: I will use these two methods to implement and compare performance, tomorrow or the day after tomorrow.

Is this considered premature optimization? For me, copying as many bytes as possible at once is always better, without considering the complexity introduced.
For me, this is "best practice".

Comment (tustvold) [reply arrow-select/src/filter.rs:745]: > copying as many bytes as possible at once is always better, without considering the complexity introduced.

Unless I am misunderstanding the only difference is eliding a capacity check, both methods will perform the same memcpy operations. The compiler can't coalesce writes in either case as they're from potentially non-contiguous source regions.

Comment (chloro-pn) [reply arrow-select/src/filter.rs:745]: Additionally,`push_unchecked` cannot handle slice types, so this method cannot be compiled successfully
```
let mut buffer = MutableBuffer::new(len.checked_mul(item_size).unwrap());
iterator.for_each(|item| unsafe { buffer.push_unchecked(item) });
```

Comment (tustvold) [reply arrow-select/src/filter.rs:745]: Yes as stated above it will have to use extend_from_slice instead, as push_unchecked can't be used with unsize types

Comment (chloro-pn) [reply arrow-select/src/filter.rs:745]: Of course, if consider the issue of introducing complexity, I can use `extend_from_slice` to refactor the code, regardless of the test results.

Comment (chloro-pn) [reply arrow-select/src/filter.rs:745]: However, for a single memory copy operation, the compiler can copy multiple bytes simultaneously by generating SIMD instructions.
For example, if you call memcpy once and copy four bytes, it is always better or equivalent to calling memcpy four times and copying one byte each time. Because the first scenario may be optimized by SIMD while the second scenario is not.

I am not sure if Rust's compiler has implemented this optimization just like C/C++ libc.

Comment (tustvold) [reply arrow-select/src/filter.rs:745]: I think there is some confusion, `extend_from_slice` can be called with a `&[u8]`, nobody is suggesting copying individual bytes

Comment (alamb) [reply arrow-select/src/filter.rs:745]: > I will use these two methods to implement and compare performance, tomorrow or the day after tomorrow.

Thank you @chloro-pn (also thank you for writing the benchmarks that can show it) -- I think this is the best approach so that we can see what the actual impact on performance is rather than theorizing what the compiler might / might not do

Comment (chloro-pn) [reply arrow-select/src/filter.rs:745]: My mistake...
When I say `extend_from_slice`, I always mistake it for `from_trusted_len_iter` : (
Sorry for wasting time, I will use `extend_from_slice` for refactoring.

**Comment by chloro-pn** [comment]
The baseline is `extend_from_slice` version.
According to the benchmark results, there is almost no difference in performance between these two methods
```
filter fsb with value length 5 (kept 1/2)
                        time:   [119.79 µs 119.95 µs 120.16 µs]
                        change: [-1.0017% +0.9824% +3.0183%] (p = 0.35 > 0.05)
                        No change in performance detected.
Found 13 outliers among 100 measurements (13.00%)
  3 (3.00%) high mild
  10 (10.00%) high severe

filter fsb with value length 5 high selectivity (kept 1023/1024)
                        time:   [20.973 µs 21.159 µs 21.352 µs]
                        change: [+1.7384% +2.8273% +3.8921%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) low mild

filter fsb with value length 5 low selectivity (kept 1/1024)
                        time:   [1.6726 µs 1.6808 µs 1.6892 µs]
                        change: [-12.122% -11.630% -11.143%] (p = 0.00 < 0.05)
                        Performance has improved.

filter context fsb with value length 5 (kept 1/2)
                        time:   [22.079 µs 22.143 µs 22.247 µs]
                        change: [-0.1067% +0.2270% +0.5189%] (p = 0.16 > 0.05)
                        No change in performance detected.
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe

filter context fsb with value length 5 high selectivity (kept 1023/1024)
                        time:   [22.071 µs 22.099 µs 22.152 µs]
                        change: [-0.1885% +0.1895% +0.5296%] (p = 0.34 > 0.05)
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) high mild
  6 (6.00%) high severe

filter context fsb with value length 5 low selectivity (kept 1/1024)
                        time:   [22.062 µs 22.095 µs 22.143 µs]
                        change: [+0.0099% +0.1751% +0.3568%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 9 outliers among 100 measurements (9.00%)
  2 (2.00%) high mild
  7 (7.00%) high severe

filter fsb with value length 20 (kept 1/2)
                        time:   [134.22 µs 134.38 µs 134.53 µs]
                        change: [+4.3318% +4.6516% +4.9943%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 13 outliers among 100 measurements (13.00%)
  5 (5.00%) low mild
  5 (5.00%) high mild
  3 (3.00%) high severe

filter fsb with value length 20 high selectivity (kept 1023/1024)
                        time:   [271.86 µs 273.74 µs 275.71 µs]
                        change: [+5.3890% +7.3050% +9.1749%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

filter fsb with value length 20 low selectivity (kept 1/1024)
                        time:   [1.6427 µs 1.6509 µs 1.6592 µs]
                        change: [-11.179% -10.468% -9.7402%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

filter context fsb with value length 20 (kept 1/2)
                        time:   [22.169 µs 22.255 µs 22.355 µs]
                        change: [-3.0221% -2.3515% -1.6009%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  3 (3.00%) high mild
  10 (10.00%) high severe

filter context fsb with value length 20 high selectivity (kept 1023/1024)
                        time:   [22.092 µs 22.294 µs 22.540 µs]
                        change: [+0.0301% +0.5035% +1.0283%] (p = 0.06 > 0.05)
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  3 (3.00%) high mild
  7 (7.00%) high severe

filter context fsb with value length 20 low selectivity (kept 1/1024)
                        time:   [22.078 µs 22.153 µs 22.278 µs]
                        change: [-0.9876% -0.0396% +0.7882%] (p = 0.93 > 0.05)
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  6 (6.00%) high mild
  4 (4.00%) high severe

filter fsb with value length 50 (kept 1/2)
                        time:   [420.94 µs 432.61 µs 444.20 µs]
                        change: [+1.7582% +4.1423% +6.5138%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 11 outliers among 100 measurements (11.00%)
  9 (9.00%) low mild
  2 (2.00%) high mild

filter fsb with value length 50 high selectivity (kept 1023/1024)
                        time:   [700.91 µs 713.64 µs 727.65 µs]
                        change: [-0.8821% +1.3435% +3.3318%] (p = 0.25 > 0.05)
                        No change in performance detected.
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe

filter fsb with value length 50 low selectivity (kept 1/1024)
                        time:   [1.8066 µs 1.8227 µs 1.8416 µs]
                        change: [-10.498% -9.3662% -7.8719%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe

filter context fsb with value length 50 (kept 1/2)
                        time:   [22.234 µs 22.498 µs 22.830 µs]
                        change: [-1.1997% -0.0554% +0.9810%] (p = 0.92 > 0.05)
                        No change in performance detected.
Found 14 outliers among 100 measurements (14.00%)
  3 (3.00%) high mild
  11 (11.00%) high severe

filter context fsb with value length 50 high selectivity (kept 1023/1024)
                        time:   [22.073 µs 22.094 µs 22.122 µs]
                        change: [-0.4814% -0.0013% +0.5338%] (p = 1.00 > 0.05)
                        No change in performance detected.
Found 13 outliers among 100 measurements (13.00%)
  1 (1.00%) high mild
  12 (12.00%) high severe

filter context fsb with value length 50 low selectivity (kept 1/1024)
                        time:   [22.112 µs 22.232 µs 22.437 µs]
                        change: [-0.4878% -0.0770% +0.5095%] (p = 0.78 > 0.05)
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  1 (1.00%) high mild
  9 (9.00%) high severe
```

**Review by tustvold — APPROVED** [review]

**Review by alamb — APPROVED** [review]
Very nice -- thank you @chloro-pn and @tustvold
